### ISSUE TITLE
Quarry can sometimes dig above or below limits

### ISSUE THREAD

**Issue by OgelGames** [mt-mods/technic#143]
From in-game chat:

### 2020/09/27
> SwissalpS: hey, quarry is digging deeper than 100!
> chi: it it after a jump?
> SwissalpS: no
> chi: i heard of some cases of digging ridiculously above or below the machine
> SwissalpS: it's really strange. I had put only a couple of nodes thick blob under quarry
> SwissalpS: and it dug 101 m under them away
> SwissalpS: well almost, it left some little shelves
> SwissalpS: I restarted them again and now it's clearing out those remains
> SwissalpS: digging at -121 now :p
> SwissalpS: oh, one is digging at -195 now :p
> SX: have any strange stuff around it?
> SwissalpS: well, I recently added digiline control
> SwissalpS: otherwise same as since months ago
> SX: from code I see only one clear way how this could happen: if dig_index : 0 and dig_index < dig_steps then if it gets somehow stuck into digging partial layers
> chi: still leaves things here to
> SX: so it never properly finishes layer but moves to next layer without finishing previous
> SX: while it is digging layer it does not check depth at all, only when switching layers
> SwissalpS: well, on this last run it was all air. only after 100m there was something to dig

### 2020/12/22
> SwissalpS: whoa, this quarry says: digging 7766m above machine
> Ruggila: did you jump while the quarry was on?
> SwissalpS: no, I have not jumped this ship in over five months
> SwissalpS: haven't dug anything with it either
> SwissalpS: when I last left it, all quarries had finished
> SwissalpS: seems some lbm got it confused. -7766 is the altitude where the quarries are. two of them had this value
> SwissalpS: others just said they were digging at normal values. still, I expected all to say: digging finished
> SX: dig level is stored in metadata and get_int returns 0 if not set. that might be reason
> SX: if dig level is absolute y
> SX: using just get can be used to verify as it returns nil for keys that have no value

**Comment by BuckarooBanzay** [comment]
Maybe there's an issue with stale metadata for the `dig_level` value:

https://github.com/mt-mods/technic/blob/c17c2e543cde2187bb892856489130e3b455ccc6/technic/machines/HV/quarry.lua#L184

**Comment by OgelGames** [comment]
Turns out this is actually two bugs in one...

The first bug actually allows the quarry to dig lower that it is supposed to, under the right conditions. If a quarry is in "looking for ground" mode, there are no depth checks, so if there is only air or vacuum (and the area is loaded), then the quarry will keep going down (past it's limit) until it finds something that isn't air or vacuum.

I was able to replicate this by placing a quarry in the sky, while keeping the area loaded with anchors, but it would also be possible if a player is at the level the quarry is searching at.

![screenshot_20201224_172921](https://user-images.githubusercontent.com/48543043/103072870-92bebd80-461a-11eb-98a6-c5e67ec6e9a0.png)

The other bug is only visual, happening only with old quarries that don't have the metadata needed to correctly show where they are supposedly digging, and are also not getting enough power to run (which would fix the values).

I confirmed this by looking at the metadata of @SwissalpS's quarries, which was clearly missing many of the metadata values a new quarry has:

Old quarry:
```
meta.fields = {
          HV_EU_demand = "10000",
          size = "8",
          dug = "30056",
          enabled = "1",
          owner = "SwissalpS",
          purge_on = "0",
          formspec = "size[8,9] ... <removed> ... label[0,4.1;Digging 7766 m above machine]",
          HV_EU_timeout = "0",
          infotext = "HV Quarry Has No Network",
          HV_network = "(-16389,-7765,-4)"
     }
```
New/restarted quarry:
```
meta.fields = {
          quarry_pos = "(-16388,-7766,-4)",
          size = "8",
          dig_steps = "289",
          dig_level = "-7763",
          owner = "SwissalpS",
          purge_on = "0",
          dug = "0",
          dig_index = "0",
          finished = "0",
          formspec = "size[8,9] ... < removed > ... label[0,4.1;Digging not started]",
          HV_EU_timeout = "2",
          infotext = "HV Quarry Disabled",
          HV_network = "1.407219470336e+14",
          quarry_dir = "1"
     }
```

**Comment by SwissalpS** [comment]
Quarries still leave traces behind. So my general move is to run them twice. This also triggers them to keep digging deeper and deeper.
I often restart them and then go far away to tend to something else. So the area the quarries are searching should not be active. (I don't think I have that many visitors or I'd run into them more often when I return)

**Comment by dennisjenkins75** [comment]
I build a brand new underground quarry on pandorabox.io.  I have a ring of 12 quarries at Y=-1604, with digiline enabled.  They have dug out the entire normal volume, down to 100 nodes below the quarries, then stopped.  I then used a node-replacer to replace the floor w/ "ultrawhite plastic" (to make the bottom visible).  I then send the quarries a digiline message to "restart", and they dug down to Y=Q.y - 101 and ate the floor, then stopped at 101 nodes below.  I then repaired the floor at depth 102 and restarted the quarries.  They then ate the floor at Y=Q.y - 102.

This matches OgelGames' observation that the quarry is not performing depth checks while looking for a diggable node.
